src/viewPanel/sessionWidget.ts

```diff
@@ -108,34 +79,67 @@ export class SessionWidget extends BoxPanel {
-    const dataPaths: { [k: string]: string } = {};
-    if ('LoadLog' in this._model.contents) {
-      const path = (this._model.contents['LoadLog'] as unknown as ILoadLog)
-        .path;
-      dataPaths[PathExt.basename(path).replace(PathExt.extname(path), '')] =
-        PathExt.join(PathExt.dirname(sessionPath), path);
-    }
-    let i = 0;
-    while (`LoadLog_${i}` in this._model.contents) {
-      const path = (this._model.contents[`LoadLog_${i}`] as unknown as ILoadLog)
-        .path;
-      dataPaths[PathExt.basename(path).replace(PathExt.extname(path), '')] =
-        PathExt.join(PathExt.dirname(sessionPath), path);
-      i++;
-    }
+    const code = `
+    import json
+    import warnings
+    from pathlib import Path
 
-    if (!dataPaths) {
-      return;
-    }
+    import y_py as Y
+    import glue_jupyter as gj
+    from ypywidgets import Widget, reactive
 
-    // TODO Handle loading errors and report in the UI?
-    const code = `
-    data_paths = json.loads('${JSON.stringify(dataPaths)}')
+    # Ignoring warnings so they don't show up in viewers
+    # TODO don't suppress warnings but redirect them?
+    warnings.filterwarnings("ignore")
+
+    class GlueSession(Widget):
+        def __init__(self, app, path: str):
+            self._app = app
+            self._path = Path(path)
+            self._data = {}
+
+            ydoc = Y.YDoc()
+            self._contents = ydoc.get_map("contents")
+            self._dataset = ydoc.get_map("dataset")
+            self._links = ydoc.get_map("links")
+            self._tabs = ydoc.get_map("tabs")
+
+            self._contents.observe(self._watch_contents)
+
+            super().__init__(
+                ydoc=ydoc,
+                comm_metadata=dict(
+                    ymodel_name="GlueSession",
+                    create_ydoc=False,
+                    path=path,
+                ),
+            )
+
+        @property
+        def data(self):
+            return self._data
+
+        def _watch_contents(self, event):
+            # extract paths to datasets
+            if "LoadLog" in event.target:
+                path = Path(event.target["LoadLog"]["path"])
+                data_paths = {}
+                data_paths[path.stem] = str(self._path.parent / path.name)
+                i = 0
+                while True:
+                    k = f"LoadLog_{i}"
+                    if k not in event.target:
+                        break
+                    path = Path(event.target[k]["path"])
+                    data_paths[path.stem] = str(self._path.parent / path.name)
+                    data_paths[path.stem] = str(self._path.parent / path.name)
+                    i += 1
+                for name, path in data_paths.items():
+                    self._data[name] = self._app.load_data(path)
+
+    app = gj.jglue()
 
-    data = {}
-    for name, path in data_paths.items():
-        data[name] = app.load_data(path)
+    glue_session = GlueSession(app, "${sessionPath}")
```

Comment:
Data is not loaded at this point, so `this._dataLoaded.resolve()` should not be called yet.

Reply:
We probably need to resolve this before merging. Otherwise plots will show up half of the times.

Maybe the `dataLoaded` promise delegate should be remove there, each tab view's Python code (the one responsible for creating the scatter/image plots) should wait for the data to be loaded Python side?

Reply:
Yes, if we are going to move more logic to the kernel then we might not need for the frontend to wait for the data to be loaded anyway. The kernel, after loading the data in the glue app in reaction to the session change, will just plot the data.

Reply:
> will just plot the data.
this is the hard part, the kernel can not plot the data itself, we might need to create an empty plot first then the kernel will populate the data for this plot later.